src/cdomains/intDomain.ml

```diff
@@ -1825,9 +1919,468 @@ module Enums : S with type int_t = BigInt.t = struct
+    | Some (c, m), Some (c', m') when (Cil.isSigned ik) || c <: Ints_t.zero || c' <: Ints_t.zero -> top()
+    | Some (c, m), Some (c', m') -> if (m =: Ints_t.zero && m' =: Ints_t.zero)
+        then Some (Ints_t.bitand (max_int ik) (Ints_t.shift_left c (Ints_t.to_int c')), Ints_t.zero)
+      else let x = (Ints_t.bitand (max_int ik) (Ints_t.shift_left Ints_t.one (Ints_t.to_int c'))) in   (* 2^c' *)
+        if is_prime (m' +: Ints_t.one) then Some (x *: c, gcd (x *: m) ((c *: x) *: (m' +: Ints_t.one)))
+          else Some (x *: c, gcd (x *: m) (c *: x))
+
+  let shift_left ik x y =
+    let res = shift_left ik x y in
+    if M.tracing then  M.trace "congruence" "shift_left : %a %a becomes %a \n" pretty x pretty y pretty res;
+    res
+
+  let mul ?(no_ov=false) ik x y =
+    match x, y with
+    | None, None -> bot ()
+    | None, _ | _, None ->
+       raise (ArithmeticOnIntegerBot (Printf.sprintf "%s op %s" (show x) (show y)))
+    | Some (c1, m1), Some (c2, m2) when no_ov ->
+       Some (c1 *: c2, gcd (c1 *: m2) (gcd (m1 *: c2) (m1 *: m2)))
+    | Some (c1, m1), Some (c2, m2)
+         when m1 =: Ints_t.zero && m2 =: Ints_t.zero && not (Cil.isSigned ik) ->
+       Some((c1 *: c2) %: ((max_int ik) +: Ints_t.one), Ints_t.zero)
+    | _ -> top ()
+
+  let mul ?no_ov ik x y =
+    let res = mul ?no_ov ik x y in
+    if M.tracing then  M.trace "congruence" "mul : %a %a -> %a \n" pretty x pretty y pretty res;
+    res
+
+  let neg ?(no_ov=false) ik x =
+    match x with
+    | None -> bot()
+    | Some _ ->  mul ~no_ov ik (of_int ik (Ints_t.of_int (-1))) x
+
+
+  let add ?(no_ov=false) ik x y =
+    match (x, y) with
+    | None, None -> bot ()
+    | None, _ | _, None ->
+       raise (ArithmeticOnIntegerBot (Printf.sprintf "%s op %s" (show x) (show y)))
+    | Some (c1, m1), Some (c2, m2) when no_ov -> normalize (Some (c1 +: c2, gcd m1 m2))
+    | Some (c1, m1), Some (c2, m2)
+         when m1 =: Ints_t.zero && m2 =: Ints_t.zero && not (Cil.isSigned ik) ->
+       Some((c1 +: c2) %: ((max_int ik) +: Ints_t.one), Ints_t.zero)
+    | _ -> top ()
+
+
+  let add ?no_ov ik x y =
+    let res = add ?no_ov ik x y in
+    if M.tracing then
+      M.trace "congruence" "add : %a %a -> %a \n" pretty x pretty y
+        pretty res ;
+    res
+
+  let sub ?(no_ov=false) ik x y = add ~no_ov ik x (neg ~no_ov ik y)
+
+
+  let sub ?no_ov ik x y =
+    let res = sub ?no_ov ik x y in
+    if M.tracing then
+      M.trace "congruence" "sub : %a %a -> %a \n" pretty x pretty y
+        pretty res ;
+    res
+
+  let bitnot ik x = match x with
+    | None -> None
+    | Some (c, m) -> if (Cil.isSigned ik) then sub ik (neg ik x) one else Some (Ints_t.sub (max_int ik) c, m)
+
+  (** The implementation of the bit operations is based on the master’s thesis
+      'Abstract Interpretation and Abstract Domains' written by Stefan Bygde.
+      see: https://www.dsi.unive.it/~avp/domains.pdf *)
```

Comment:
This is missing the weakening operation ξ from the paper, and hence the
`Ints_t.shift_left Ints_t.one (Ints_t.to_int (min m m')))` can be by ridiculous amounts which cause the program to run out of memory.